be/src/exec/spill/spill_components.cpp

```diff
@@ -0,0 +1,352 @@
+}
+
+Status RawSpillerWriter::prepare(RuntimeState* state) {
+    if (!_mem_table_pool.empty()) {
+        return Status::OK();
+    }
+
+    const auto& opts = options();
+    for (size_t i = 0; i < opts.mem_table_pool_size; ++i) {
+        if (opts.is_unordered) {
+            _mem_table_pool.push(std::make_unique<UnorderedMemTable>(state, opts.spill_file_size, _parent_tracker));
+        } else {
+            _mem_table_pool.push(std::make_unique<OrderedMemTable>(&opts.sort_exprs->lhs_ordering_expr_ctxs(),
+                                                                   opts.sort_desc, state, opts.spill_file_size,
+                                                                   _parent_tracker));
+        }
+    }
+
+    return Status::OK();
+}
+
+Status RawSpillerWriter::flush_task(RuntimeState* state, const MemTablePtr& mem_table) {
+    if (state->is_cancelled()) {
+        return Status::OK();
+    }
+
+    const auto& serde = _spiller->serde();
+    spill::AcquireBlockOptions opts;
+    opts.query_id = state->query_id();
+    opts.plan_node_id = options().plan_node_id;
+    opts.name = options().name;
+    ASSIGN_OR_RETURN(auto block, _spiller->block_manager()->acquire_block(opts));
+
+    // TODO: reuse io context
+    SerdeContext spill_ctx;
+    // DCHECK(writable != nullptr);
+    {
+        SCOPED_TIMER(_spiller->metrics().write_io_timer);
+        // flush all pending result to spilled files
+        size_t num_rows_flushed = 0;
+        RETURN_IF_ERROR(mem_table->flush([&](const auto& chunk) {
+            num_rows_flushed += chunk->num_rows();
+            RETURN_IF_ERROR(serde->serialize(spill_ctx, chunk, block));
+            return Status::OK();
+        }));
+        TRACE_SPILL_LOG << "spill flush rows:" << num_rows_flushed << ",spiller:" << this;
+    }
+
+    // be careful close method return a not ok status
+    // then release the pending memory
+    // flush
+    RETURN_IF_ERROR(block->flush());
+    _spiller->block_manager()->release_block(block);
+
+    {
+        std::lock_guard<std::mutex> l(_mutex);
+        _block_group.append(std::move(block));
+    }
+
+    return Status::OK();
+}
+
+Status RawSpillerWriter::acquire_stream(const SpillPartitionInfo* partition,
+                                        std::shared_ptr<SpillInputStream>* stream) {
+    return acquire_stream(stream);
+}
+
+Status RawSpillerWriter::acquire_stream(std::shared_ptr<SpillInputStream>* stream) {
+    std::shared_ptr<SpillInputStream> input_stream;
+    const auto& serde = _spiller->serde();
+    const auto& opts = options();
+
+    if (opts.is_unordered) {
+        ASSIGN_OR_RETURN(input_stream, _block_group.as_unordered_stream(serde));
+    } else {
+        ASSIGN_OR_RETURN(input_stream,
+                         _block_group.as_ordered_stream(_runtime_state, serde, opts.sort_exprs, opts.sort_desc));
+    }
+
+    *stream = input_stream;
+
+    if (_mem_table != nullptr && !_mem_table->is_empty()) {
```

Comment:
is there a thread-safety issue? 
1. try to read _mem_table during appending chunk to it.
2. _mem_table is flushed to a block but not removed in memory,  we count the block and the MemTable twice.

Reply:
1. acquire_stream is only called when there are no any in-coming writes to the partition
2. if mem_table flushed into a block. it will also be released to pool.